zstd/zstd.go

```diff
@@ -38,161 +35,91 @@ func (c *CompressionCodec) Name() string { return "zstd" }
 
 // NewReader implements the kafka.CompressionCodec interface.
 func (c *CompressionCodec) NewReader(r io.Reader) io.ReadCloser {
-	return &reader{
-		reader: r,
-		buffer: bufferPool.Get().(*buffer),
+	p := new(reader)
+	if cached := decPool.Get(); cached == nil {
```

Comment:
I think if we set the `New` function on the pool this condition becomes impossible (`nil` will never be returned).

You may have found the error returned by `zstd.NewReader` challenging to handle in the context of the pool constructor, see how we've solved that in the [`kafka-go/gzip`](https://github.com/segmentio/kafka-go/blob/master/gzip/gzip.go#L13-L31) case, I think you can mirror it here.

Reply:
That is exactly why I don't set the New function. The `nil` value is documented behaviour of sync.Pool.